presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java

```diff
@@ -0,0 +1,192 @@
+        Class<? extends AccumulatorState> stateInterface = BigIntegerAndLongState.class;
+        AccumulatorStateSerializer<?> stateSerializer = new BigIntegerAndLongStateSerializer();
+
+        if (type.getJavaType() == long.class) {
+            inputFunction = SHORT_DECIMAL_INPUT_FUNCTION;
+            outputFunction = SHORT_DECIMAL_OUTPUT_FUNCTION;
+            combineFunction = COMBINE_FUNCTION;
+        }
+        else {
+            inputFunction = LONG_DECIMAL_INPUT_FUNCTION;
+            outputFunction = LONG_DECIMAL_OUTPUT_FUNCTION;
+            combineFunction = COMBINE_FUNCTION;
+        }
+        inputFunction = inputFunction.bindTo(type);
+        outputFunction = outputFunction.bindTo(type);
+
+        AggregationMetadata metadata = new AggregationMetadata(
+                generateAggregationName(NAME, type, inputTypes),
+                createInputParameterMetadata(type),
+                inputFunction,
+                null,
+                null,
+                combineFunction,
+                outputFunction,
+                stateInterface,
+                stateSerializer,
+                new BigIntegerAndLongStateFactory(),
+                type,
+                false);
+
+        Type intermediateType = stateSerializer.getSerializedType();
+        GenericAccumulatorFactoryBinder factory = new AccumulatorCompiler().generateAccumulatorFactoryBinder(metadata, classLoader);
+        return new InternalAggregationFunction(NAME, inputTypes, intermediateType, type, true, false, factory);
+    }
+
+    private static List<ParameterMetadata> createInputParameterMetadata(Type type)
+    {
+        return ImmutableList.of(new ParameterMetadata(STATE), new ParameterMetadata(BLOCK_INPUT_CHANNEL, type), new ParameterMetadata(BLOCK_INDEX));
+    }
+
+    public static void inputShortDecimal(Type type, BigIntegerAndLongState state, Block block, int position)
+    {
+        initializeIfNeeded(state);
+        BigInteger value = BigInteger.valueOf(type.getLong(block, position));
+        state.setBigInteger(state.getBigInteger().add(value));
+        state.setLong(state.getLong() + 1);
+    }
+
+    public static void inputLongDecimal(Type type, BigIntegerAndLongState state, Block block, int position)
+    {
+        initializeIfNeeded(state);
+        BigInteger value = unscaledValueToBigInteger(type.getSlice(block, position));
+        BigInteger accumulated = state.getBigInteger();
+        state.setBigInteger(accumulated.add(value));
+        state.setLong(state.getLong() + 1);
+    }
+
+    private static void initializeIfNeeded(BigIntegerAndLongState state)
+    {
+        if (state.getBigInteger() == null) {
+            state.setBigInteger(BigInteger.valueOf(0));
+        }
+    }
+
+    public static void combine(BigIntegerAndLongState state, BigIntegerAndLongState otherState)
+    {
+        state.setLong(state.getLong() + otherState.getLong());
+        if (state.getBigInteger() == null) {
```

Comment:
`state` is always non-null, null can be the properties inside it. That's a soft assumption based on looking at other classes and how they interact with `state`s